File: frontend/src/services/achiever.ts
import apiClient from './apiClient';

export interface AchieverContext {
  student: {
    full_name: string | null;
    phone: string | null;
    location: string | null;
    bio: string | null;
  };
  academic: {
    degree: string | null;
    branch: string | null;
    academic_year: string | null;
    graduation_year: number | null;
  };
  professional: {
    headline: string | null;
    current_status: string | null;
    experience_level: string | null;
    years_of_experience: number | null;
    professional_summary: string | null;
    work_domain: string | null;
  };
  target: {
    target_company: { id: number; name: string; description: string | null } | null;
    target_role: { id: number; name: string; description: string | null } | null;
    career_role: { id: number; name: string; description: string | null } | null;
  };
  external_profiles: {
    github_status: string;
    github_url: string | null;
    linkedin_status: string;
    linkedin_url: string | null;
    portfolio_status: string;
    portfolio_url: string | null;
  };
  resume: {
    filename: string | null;
    mime_type: string | null;
    file_size: number | null;
    upload_date: string | null;
    active_status: boolean;
    extraction_status: string;
    extracted_text: string | null;
  };
  evidence: Array<{
    type: string;
    status: string;
    title: string | null;
    metadata: any;
    created_at: string | null;
    updated_at: string | null;
  }>;
  skills: Array<{
    skill: { id: number; name: string; category: { id: number; name: string } };
    proficiency_level: number;
    source_status: string | null;
    updated_timestamp: string | null;
  }>;
  skill_gaps: Array<{
    skill: { id: number; name: string; category: { id: number; name: string } };
    required_proficiency: number;
    current_proficiency: number;
    status: string;
    requirement_source: string;
  }>;
  requirements: Array<{
    skill: { id: number; name: string; category: { id: number; name: string } };
    required_proficiency: number;
    source: string;
  }>;
  metadata: {
    context_version: string;
    generated_at: string;
    requirement_source: string;
    available_evidence_count: number;
    verified_evidence_count: number;
    unverified_evidence_count: number;
    has_resume: boolean;
    has_resume_text: boolean;
    has_github: boolean;
    has_linkedin: boolean;
    has_portfolio: boolean;
    has_assessment: boolean;
    has_target_company: boolean;
    has_target_role: boolean;
  };
}

export const getAchieverContext = async (): Promise<AchieverContext> => {
  const response = await apiClient.get<AchieverContext>('/api/v1/student/achiever/context');
  return response.data;
};

export interface EvidenceSummaryItem {
  source: string;
  status: string;
}

export interface PrioritizedSkillGap {
  skill: { id: number; name: string; category: { id: number; name: string } };
  required_proficiency: number;
  current_proficiency: number;
  gap_status: string;
  priority: string;
  requirement_source: string;
  evidence_summary: EvidenceSummaryItem[];
  dependency_status: string;
  reasons: string[];
}

export interface AchieverPriorityResponse {
  target: any; // We can type this fully if needed, same as in AchieverContext
  requirement_source: string;
  prioritized_gaps: PrioritizedSkillGap[];
  summary: string;
}

export const getAchieverPriorities = async (): Promise<AchieverPriorityResponse> => {
  const response = await apiClient.get<AchieverPriorityResponse>('/api/v1/student/achiever/priorities');
  return response.data;
};

export interface KnowledgeResourceResponse {
  resource_id: number;
  title: string;
  resource_type: string;
  provider: string;
  difficulty: string | null;
  source_url: string;
  estimated_duration: string | null;
  similarity_score: number;
}

export interface SkillResourceGroup {
  skill_id: number;
  skill_name: string;
  priority: string;
  resources: KnowledgeResourceResponse[];
}

export interface AchieverResourcesResponse {
  student_id: number;
  target: {
    company: string;
    role: string;
  };
  resources: SkillResourceGroup[];
}

export const getAchieverResources = async (): Promise<AchieverResourcesResponse> => {
  const response = await apiClient.get<AchieverResourcesResponse>('/api/v1/student/achiever/resources');
  return response.data;
};

export interface RoadmapResource {
  resource_id: number;
  title: string;
  source_url: string;
  difficulty?: string;
  provider?: string;
}

export interface RoadmapStep {
  id: number;
  roadmap_id: number;
  step_order: number;
  skill_id: number;
  skill_name: string;
  priority: string;
  objective: string;
  description?: string;
  estimated_duration?: string;
  resources: RoadmapResource[];
  practice_task?: string;
  status: string;
}

export interface RoadmapResponse {
  roadmap_id: number;
  target: {
    company?: string;
    role?: string;
  };
  summary?: string;
  status: string;
  steps: RoadmapStep[];
}

export const getRoadmap = async (): Promise<RoadmapResponse | null> => {
  try {
    const response = await apiClient.get<RoadmapResponse>('/api/v1/student/achiever/roadmap');
    return response.data;
  } catch (error: any) {
    if (error.response && error.response.status === 404) {
      return null;
    }
    throw error;
  }
};

export const generateRoadmap = async (): Promise<RoadmapResponse> => {
  const response = await apiClient.post<RoadmapResponse>('/api/v1/student/achiever/roadmap/generate');
  return response.data;
};

File: backend/app/schemas/achiever.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List, Dict, Any
from datetime import datetime
from app.schemas.skill import SkillResponse

class AchieverStudentContext(BaseModel):
    full_name: Optional[str] = None
    phone: Optional[str] = None
    location: Optional[str] = None
    bio: Optional[str] = None

class AchieverAcademicContext(BaseModel):
    degree: Optional[str] = None
    branch: Optional[str] = None
    academic_year: Optional[str] = None
    graduation_year: Optional[int] = None

class AchieverProfessionalContext(BaseModel):
    headline: Optional[str] = None
    current_status: Optional[str] = None
    experience_level: Optional[str] = None
    years_of_experience: Optional[float] = None
    professional_summary: Optional[str] = None
    work_domain: Optional[str] = None

class AchieverTargetCompanyRoleContext(BaseModel):
    id: int
    name: str
    description: Optional[str] = None

class AchieverTargetContext(BaseModel):
    target_company: Optional[AchieverTargetCompanyRoleContext] = None
    target_role: Optional[AchieverTargetCompanyRoleContext] = None
    career_role: Optional[AchieverTargetCompanyRoleContext] = None

class AchieverExternalProfilesContext(BaseModel):
    github_status: str
    github_url: Optional[str] = None
    linkedin_status: str
    linkedin_url: Optional[str] = None
    portfolio_status: str
    portfolio_url: Optional[str] = None

class AchieverResumeContext(BaseModel):
    filename: Optional[str] = None
    mime_type: Optional[str] = None
    file_size: Optional[int] = None
    upload_date: Optional[datetime] = None
    active_status: bool = False
    extraction_status: str
    extracted_text: Optional[str] = None

class AchieverEvidenceContext(BaseModel):
    type: str
    status: str
    title: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

class AchieverSkillContext(BaseModel):
    skill: SkillResponse
    proficiency_level: int
    source_status: Optional[str] = None
    updated_timestamp: Optional[datetime] = None

class AchieverRequirementContext(BaseModel):
    skill: SkillResponse
    required_proficiency: int
    source: str

class AchieverSkillGapContext(BaseModel):
    skill: SkillResponse
    required_proficiency: int
    current_proficiency: int
    status: str
    requirement_source: str

class AchieverMetadataContext(BaseModel):
    context_version: str
    generated_at: datetime
    requirement_source: str
    available_evidence_count: int
    verified_evidence_count: int
    unverified_evidence_count: int
    has_resume: bool
    has_resume_text: bool
    has_github: bool
    has_linkedin: bool
    has_portfolio: bool
    has_assessment: bool
    has_target_company: bool
    has_target_role: bool

class AchieverContextResponse(BaseModel):
    student: AchieverStudentContext
    academic: AchieverAcademicContext
    professional: AchieverProfessionalContext
    target: AchieverTargetContext
    external_profiles: AchieverExternalProfilesContext
    resume: AchieverResumeContext
    evidence: List[AchieverEvidenceContext]
    skills: List[AchieverSkillContext]
    skill_gaps: List[AchieverSkillGapContext]
    requirements: List[AchieverRequirementContext]
    metadata: AchieverMetadataContext

class EvidenceSummaryItem(BaseModel):
    source: str
    status: str

class PrioritizedSkillGap(BaseModel):
    skill: SkillResponse
    required_proficiency: int
    current_proficiency: int
    gap_status: str
    priority: str
    requirement_source: str
    evidence_summary: List[EvidenceSummaryItem]
    dependency_status: str
    reasons: List[str]

class AchieverPriorityResponse(BaseModel):
    target: AchieverTargetContext
    requirement_source: str
    prioritized_gaps: List[PrioritizedSkillGap]
    summary: str
